### PR TITLE
[DESIGN] Exception handling proposal

### PR DESCRIPTION
**Pull request by dennisseah**
related to https://github.com/microsoft/bedrock/issues/1099

### CODE REVIEW

**Review by andrebriggs — APPROVED** [review]

File: technical-docs/designs/exceptionHandling.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+```
+try {
+  ...
+} catch (err) {
+  logger.error(err);
+}
+```
+
+Reader of the code has a hard time understand why error is ignored.
```
Comment (yradsmikham) [inline technical-docs/designs/exceptionHandling.md]: "understanding" ..?

File: technical-docs/designs/exceptionHandling.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+## 2. Out of Scope
+
+This document shall not cover exceptions/errors that are thrown by third parties
+software and/or from Microsoft's client API (e.g. from Azure DevOps).
```
Comment (yradsmikham) [inline technical-docs/designs/exceptionHandling.md]: "third party software"?

File: technical-docs/designs/exceptionHandling.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    // using this call to check if project exist or not.
+    logger.info(`Unable to get project, ${name}`);
+  }
+  return false;
+```
+
+The proposal is to have adequate comments in the code to explain why exception
+is caught and ignored. Typically, we do not advise people to do this unless we
+are sure about it. Taking the example above, we can authentication and
+authorization failures too.
```
Comment (yradsmikham) [inline technical-docs/designs/exceptionHandling.md]: "authenticate" and "authorize"?

File: technical-docs/designs/exceptionHandling.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+Most of the time, we have a good idea of issues when we have the complete
+exception chain. For instance
+
+```
+ o Command excecute function calls
+   o Pipeline service helper function which calls
+     o Azdo pipeline API via web client
+```
+
+The proposal is to have a way to maintain this exception chain; and preserving
```
Comment (yradsmikham) [inline technical-docs/designs/exceptionHandling.md]: "preserve"

File: technical-docs/designs/exceptionHandling.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    "message": "Execute of installPipeline function failed."
+    "error": {
+      "statusCode": 401,
+      "message": "Unauthorized access to pipeline...."
+    }
+  }
+}
+```
+
+And this JSON shall be `stringified` and logged as `error` in the finally output
```
Comment (yradsmikham) [inline technical-docs/designs/exceptionHandling.md]: "final"

**Review by yradsmikham — CHANGES_REQUESTED** [review]

File: technical-docs/designs/exceptionHandling.md (hunk 1 above)
Comment (dennisseah) [reply technical-docs/designs/exceptionHandling.md]: accept

File: technical-docs/designs/exceptionHandling.md (hunk 2 above)
Comment (dennisseah) [reply technical-docs/designs/exceptionHandling.md]: accept

File: technical-docs/designs/exceptionHandling.md (hunk 3 above)
Comment (dennisseah) [reply technical-docs/designs/exceptionHandling.md]: accept

File: technical-docs/designs/exceptionHandling.md (hunk 4 above)
Comment (dennisseah) [reply technical-docs/designs/exceptionHandling.md]: accept

File: technical-docs/designs/exceptionHandling.md (hunk 5 above)
Comment (dennisseah) [reply technical-docs/designs/exceptionHandling.md]: accept.
thanks

**Review by mtarng — APPROVED** [review]

**Review by yradsmikham — APPROVED** [review]

**Pull request merged**